File: frontend/js/advanced.js
/**
 * Advanced ML Features - Frontend JavaScript
 * Fine-tuned BioBERT, Medication NER, Anomaly Detection, Longitudinal Tracking
 */

class AdvancedML {
    constructor() {
        this.currentAnalysisId = null;
        this.init();
    }

    init() {
        console.log('Advanced ML Features initialized');
        this.checkServiceHealth();
    }

    // =================================================================
    // SERVICE HEALTH CHECK
    // =================================================================
    
    async checkServiceHealth() {
        try {
            const response = await apiRequest('/advanced/health-check', {
                method: 'GET'
            });
            
            console.log('Advanced ML Service Status:', response);
            
            // Show feature availability
            this.updateFeatureAvailability(response);
            
        } catch (error) {
            console.error('Advanced ML service check failed:', error);
        }
    }

    updateFeatureAvailability(status) {
        // Update UI based on available features
        const features = {
            'fine_tuned_biobert': status.fine_tuned_biobert,
            'medication_ner': status.medication_ner,
            'anomaly_detector': status.anomaly_detector,
            'longitudinal_tracking': status.longitudinal_tracking
        };

        console.log('Available Advanced Features:', features);
    }

    // =================================================================
    // 1. FINE-TUNED BIOBERT ENTITY EXTRACTION
    // =================================================================
    
    async extractEntitiesAdvanced(analysisId) {
        try {
            console.log('Extracting entities for analysis:', analysisId);
            
            const response = await apiRequest(`/advanced/entity-extraction/${analysisId}`, {
                method: 'POST'
            });
            
            console.log('Entity extraction response:', response);
            
            if (response.entities) {
                this.displayAdvancedEntities(response);
            }
            
            return response;
            
        } catch (error) {
            console.error('Entity extraction failed:', error);
            showToast('Failed to extract entities', 'error');
            throw error;
        }
    }

    displayAdvancedEntities(data) {
        const container = document.getElementById('advancedEntities');
        if (!container) return;

        const { entities, entity_count } = data;
        
        let html = `
            <div class="advanced-feature-card">
                <div class="feature-header">
                    <h3>🧬 Advanced Entity Recognition</h3>
                    <span class="badge badge-success">${entity_count} entities found</span>
                </div>
                <p class="feature-description">Using fine-tuned BioBERT for medical entity extraction</p>
        `;

        // Diseases
        if (entities.diseases && entities.diseases.length > 0) {
            html += `
                <div class="entity-group">
                    <h4>🦠 Diseases & Conditions (${entities.diseases.length})</h4>
                    <div class="entity-list">
                        ${entities.diseases.map(e => `
                            <div class="entity-badge">
                                <span class="entity-text">${e.text}</span>
                                <span class="entity-score">${(e.score * 100).toFixed(1)}%</span>
                            </div>
                        `).join('')}
                    </div>
                </div>
            `;
        }

        // Chemicals/Drugs
        if (entities.chemicals && entities.chemicals.length > 0) {
            html += `
                <div class="entity-group">
                    <h4>💊 Chemicals & Drugs (${entities.chemicals.length})</h4>
                    <div class="entity-list">
                        ${entities.chemicals.map(e => `
                            <div class="entity-badge">
                                <span class="entity-text">${e.text}</span>
                                <span class="entity-score">${(e.score * 100).toFixed(1)}%</span>
                            </div>
                        `).join('')}
                    </div>
                </div>
            `;
        }

        // Genes
        if (entities.genes && entities.genes.length > 0) {
            html += `
                <div class="entity-group">
                    <h4>🧬 Genes (${entities.genes.length})</h4>
                    <div class="entity-list">
                        ${entities.genes.map(e => `
                            <div class="entity-badge">
                                <span class="entity-text">${e.text}</span>
                                <span class="entity-score">${(e.score * 100).toFixed(1)}%</span>
                            </div>
                        `).join('')}
                    </div>
                </div>
            `;
        }

        // Proteins
        if (entities.proteins && entities.proteins.length > 0) {
            html += `
                <div class="entity-group">
                    <h4>🧪 Proteins (${entities.proteins.length})</h4>
                    <div class="entity-list">
                        ${entities.proteins.map(e => `
                            <div class="entity-badge">
                                <span class="entity-text">${e.text}</span>
                                <span class="entity-score">${(e.score * 100).toFixed(1)}%</span>
                            </div>
                        `).join('')}
                    </div>
                </div>
            `;
        }

        html += `</div>`;
        container.innerHTML = html;
        container.style.display = 'block';
    }

    // =================================================================
    // 2. CUSTOM MEDICATION NER
    // =================================================================
    
    async extractMedications(analysisId) {
        try {
            console.log('Extracting medications for analysis:', analysisId);
            
            const response = await apiRequest(`/advanced/medication-extraction/${analysisId}`, {
                method: 'POST'
            });
            
            console.log('Medication extraction response:', response);
            
            if (response.medications) {
                this.displayMedications(response);
            }
            
            return response;
            
        } catch (error) {
            console.error('Medication extraction failed:', error);
            showToast('Failed to extract medications', 'error');
            throw error;
        }
    }

    displayMedications(data) {
        const container = document.getElementById('medications');
        if (!container) return;

        const { medications, medication_count } = data;
        
        if (medication_count === 0) {
            container.innerHTML = `
                <div class="advanced-feature-card">
                    <h3>💊 Medications</h3>
                    <p>No medications detected in this report.</p>
                </div>
            `;
            container.style.display = 'block';
            return;
        }

        let html = `
            <div class="advanced-feature-card">
                <div class="feature-header">
                    <h3>💊 Medication Analysis</h3>
                    <span class="badge badge-info">${medication_count} medications</span>
                </div>
                <p class="feature-description">Custom NER model for medication extraction</p>
                
                <div class="medication-list">
        `;

        medications.forEach(med => {
            const confidenceClass = med.confidence >= 0.75 ? 'high' : med.confidence >= 0.5 ? 'medium' : 'low';
            
            html += `
                <div class="medication-card">
                    <div class="medication-header">
                        <span class="medication-name">${med.drug_name}</span>
                        <span class="confidence-badge ${confidenceClass}">${(med.confidence * 100).toFixed(0)}%</span>
                    </div>
                    <div class="medication-details">
                        <div class="detail-item">
                            <span class="detail-label">Dosage:</span>
                            <span class="detail-value">${med.dosage}</span>
                        </div>
                        <div class="detail-item">
                            <span class="detail-label">Frequency:</span>
                            <span class="detail-value">${med.frequency}</span>
                        </div>
                    </div>
                    <div class="medication-context">${med.context}</div>
                </div>
            `;
        });

        html += `
                </div>
            </div>
        `;
        
        container.innerHTML = html;
        container.style.display = 'block';
    }

    // =================================================================
    // 3. LAB RESULT ANOMALY DETECTION
    // =================================================================
    
    async detectAnomalies(analysisId) {
        try {
            console.log('Detecting anomalies for analysis:', analysisId);
            
            const response = await apiRequest(`/advanced/anomaly-detection/${analysisId}`, {
                method: 'POST'
            });
            
            console.log('Anomaly detection response:', response);
            
            if (response.anomalies) {
                this.displayAnomalies(response);
            }
            
            return response;
            
        } catch (error) {
            console.error('Anomaly detection failed:', error);
            showToast('Failed to detect anomalies', 'error');
            throw error;
        }
    }

    displayAnomalies(data) {
        const container = document.getElementById('anomalies');
        if (!container) return;

        const { anomalies, anomaly_count, total_metrics } = data;
        
        let html = `
            <div class="advanced-feature-card anomaly-card">
                <div class="feature-header">
                    <h3>⚠️ Anomaly Detection</h3>
                    <span class="badge ${anomaly_count > 0 ? 'badge-warning' : 'badge-success'}">
                        ${anomaly_count} / ${total_metrics} anomalous
                    </span>
                </div>
                <p class="feature-description">Isolation Forest algorithm for outlier detection</p>
        `;

        if (anomaly_count === 0) {
            html += `
                <div class="no-anomalies">
                    <p>✅ All lab results within expected patterns!</p>
                </div>
            `;
        } else {
            html += `<div class="anomaly-list">`;
            
            anomalies.forEach(anomaly => {
                const severityIcon = anomaly.anomaly_severity === 'high' ? '🔴' : '🟡';
                
                html += `
                    <div class="anomaly-item ${anomaly.anomaly_severity}">
                        <div class="anomaly-header">
                            <span class="anomaly-icon">${severityIcon}</span>
                            <span class="anomaly-metric">${anomaly.metric_name}</span>
                            <span class="anomaly-value">${anomaly.metric_value} ${anomaly.unit || ''}</span>
                        </div>
                        <div class="anomaly-reason">${anomaly.anomaly_reason}</div>
                        <div class="anomaly-score">
                            Anomaly Score: <strong>${anomaly.anomaly_score.toFixed(3)}</strong>
                        </div>
                    </div>
                `;
            });
            
            html += `</div>`;
        }

        html += `</div>`;
        container.innerHTML = html;
        container.style.display = 'block';
    }

    // =================================================================
    // 4. LONGITUDINAL HEALTH TRACKING
    // =================================================================
    
    async analyzeLongitudinal() {
        try {
            console.log('Analyzing longitudinal health trends...');
            
            const response = await apiRequest('/advanced/longitudinal-analysis', {
                method: 'GET'
            });
            
            console.log('Longitudinal analysis response:', response);
            
            if (response.trends) {
                this.displayLongitudinalAnalysis(response);
            }
            
            return response;
            
        } catch (error) {
            console.error('Longitudinal analysis failed:', error);
            showToast('Failed to analyze trends', 'error');
            throw error;
        }
    }

    displayLongitudinalAnalysis(data) {
        const container = document.getElementById('longitudinalAnalysis');
        if (!container) return;

        const { trends, predictions, risk_changes, insights, data_points, tracking_period_days } = data;
        
        let html = `
            <div class="advanced-feature-card longitudinal-card">
                <div class="feature-header">
                    <h3>📈 Longitudinal Health Tracking</h3>
                    <span class="badge badge-primary">${data_points} checkups over ${tracking_period_days} days</span>
                </div>
                <p class="feature-description">Time-series analysis of your health metrics</p>
        `;

        // Insights
        if (insights && insights.length > 0) {
            html += `
                <div class="insights-section">
                    <h4>💡 Key Insights</h4>
                    <ul class="insights-list">
                        ${insights.map(insight => `<li>${insight}</li>`).join('')}
                    </ul>
                </div>
            `;
        }

        // Trends
        if (trends && trends.length > 0) {
            html += `
                <div class="trends-section">
                    <h4>📊 Metric Trends</h4>
                    <div class="trends-grid">
            `;
            
            trends.forEach(trend => {
                const directionIcon = trend.direction === 'increasing' ? '📈' : 
                                     trend.direction === 'decreasing' ? '📉' : '➡️';
                const directionClass = trend.direction;
                
                html += `
                    <div class="trend-card ${directionClass}">
                        <div class="trend-header">
                            <span class="trend-icon">${directionIcon}</span>
                            <span class="trend-name">${trend.metric_name}</span>
                        </div>
                        <div class="trend-values">
                            <div class="value-item">
                                <span class="value-label">From:</span>
                                <span class="value-number">${trend.first_value}</span>
                            </div>
                            <div class="value-item">
                                <span class="value-label">To:</span>
                                <span class="value-number">${trend.last_value}</span>
                            </div>
                        </div>
                        <div class="trend-change">
                            Change: <strong>${trend.change > 0 ? '+' : ''}${trend.change}</strong>
                            (${trend.change_percent > 0 ? '+' : ''}${trend.change_percent}%)
                        </div>
                        <div class="trend-period">
                            Over ${trend.time_span_days} days (${trend.data_points} points)
                        </div>
                    </div>
                `;
            });
            
            html += `
                    </div>
                </div>
            `;
        }

        // Predictions
        if (predictions && predictions.length > 0) {
            html += `
                <div class="predictions-section">
                    <h4>🔮 Future Predictions (30 days)</h4>
                    <div class="predictions-list">
            `;
            
            predictions.forEach(pred => {
                html += `
                    <div class="prediction-card">
                        <div class="prediction-metric">${pred.metric_name}</div>
                        <div class="prediction-values">
                            <span class="current">Current: ${pred.current_value}</span>
                            <span class="arrow">→</span>
                            <span class="predicted">Predicted: ${pred.predicted_value}</span>
                        </div>
                        <div class="prediction-confidence">
                            Confidence: <span class="badge badge-${pred.confidence}">${pred.confidence}</span>
                        </div>
                    </div>
                `;
            });
            
            html += `
                    </div>
                </div>
            `;
        }

        // Risk Changes
        if (risk_changes && risk_changes.length > 0) {
            html += `
                <div class="risk-changes-section">
                    <h4>⚠️ Risk Level Changes</h4>
                    <div class="risk-changes-list">
            `;
            
            risk_changes.forEach(risk => {
                const severityClass = risk.severity === 'high' ? 'danger' : 'warning';
                
                html += `
                    <div class="risk-change-item ${severityClass}">
                        <span class="risk-icon">⚠️</span>
                        <span class="risk-message">${risk.message}</span>
                        <span class="risk-badge badge-${severityClass}">${risk.severity}</span>
                    </div>
                `;
            });
            
            html += `
                    </div>
                </div>
            `;
        }

        html += `</div>`;
        container.innerHTML = html;
        container.style.display = 'block';
    }

    // =================================================================
    // COMBINED ADVANCED ANALYSIS
    // =================================================================
    
    async runFullAdvancedAnalysis(analysisId) {
        this.currentAnalysisId = analysisId;
        
        try {
            // Run all advanced features in parallel
            const results = await Promise.allSettled([
                this.extractEntitiesAdvanced(analysisId),
                this.extractMedications(analysisId),
                this.detectAnomalies(analysisId),
                this.analyzeLongitudinal()
            ]);

            console.log('Advanced analysis complete:', results);
            
            // Count successful analyses
            const successful = results.filter(r => r.status === 'fulfilled').length;
            
            showToast(
                `Advanced analysis complete! ${successful}/4 features executed successfully.`,
                'success'
            );
            
            return results;
            
        } catch (error) {
            console.error('Error running full advanced analysis:', error);
            showToast('Some advanced features failed', 'warning');
        }
    }
}

// Initialize global instance
const advancedML = new AdvancedML();

// Expose to window
window.advancedML = advancedML;

File: backend/app/routers/advanced.py
"""
Advanced ML Features Router
Fine-tuned BioBERT, Custom NER, Anomaly Detection, Longitudinal Tracking
"""
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Dict, Any
import logging

from app.database import get_db
from app.models import Analysis, MedicalMetric, HealthInsight, UploadedFile
from app.routers.auth import get_current_user
from app.services.advanced_ml_service import advanced_ml_service
from pydantic import BaseModel

router = APIRouter(
    prefix="/api/v1/advanced",
    tags=["advanced-ml"]
)

logger = logging.getLogger(__name__)


# ===== REQUEST/RESPONSE MODELS =====

class AdvancedEntityExtractionResponse(BaseModel):
    """Response for advanced entity extraction"""
    analysis_id: str
    entities: Dict[str, List[Dict]]
    entity_count: int
    message: str


class MedicationExtractionResponse(BaseModel):
    """Response for medication extraction"""
    analysis_id: str
    medications: List[Dict[str, Any]]
    medication_count: int
    message: str


class AnomalyDetectionResponse(BaseModel):
    """Response for anomaly detection"""
    analysis_id: str
    anomalies: List[Dict[str, Any]]
    anomaly_count: int
    total_metrics: int
    message: str


class LongitudinalAnalysisResponse(BaseModel):
    """Response for longitudinal analysis"""
    user_id: str
    trends: List[Dict[str, Any]]
    predictions: List[Dict[str, Any]]
    risk_changes: List[Dict[str, Any]]
    insights: List[str]
    data_points: int
    tracking_period_days: int
    message: str


# ===== ENDPOINTS =====

@router.post("/entity-extraction/{analysis_id}", response_model=AdvancedEntityExtractionResponse)
async def extract_entities_advanced(
    analysis_id: str,
    current_user: dict = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Extract medical entities using fine-tuned BioBERT
    More accurate than base model - detects diseases, chemicals, genes, proteins
    """
    try:
        # Get analysis
        analysis = db.query(Analysis).filter(
            Analysis.id == analysis_id
        ).first()
        
        if not analysis:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Analysis not found"
            )
        
        # Get extracted text
        text = analysis.extracted_text
        if not text:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No extracted text available"
            )
        
        # Extract entities with fine-tuned BioBERT
        entities = advanced_ml_service.extract_medical_entities_advanced(text)
        
        entity_count = sum(len(v) for v in entities.values())
        
        return AdvancedEntityExtractionResponse(
            analysis_id=analysis_id,
            entities=entities,
            entity_count=entity_count,
            message=f"Extracted {entity_count} entities using fine-tuned BioBERT"
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in advanced entity extraction: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to extract entities: {str(e)}"
        )


@router.post("/medication-extraction/{analysis_id}", response_model=MedicationExtractionResponse)
async def extract_medications_advanced(
    analysis_id: str,
    current_user: dict = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Extract medications using custom NER model
    Detects drug names, dosages, and frequencies
    """
    try:
        # Get analysis
        analysis = db.query(Analysis).filter(
            Analysis.id == analysis_id
        ).first()
        
        if not analysis:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Analysis not found"
            )
        
        # Get extracted text
        text = analysis.extracted_text
        if not text:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No extracted text available"
            )
        
        # Extract medications with custom NER
        medications = advanced_ml_service.extract_medications_advanced(text)
        
        return MedicationExtractionResponse(
            analysis_id=analysis_id,
            medications=medications,
            medication_count=len(medications),
            message=f"Extracted {len(medications)} medications with custom NER model"
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in medication extraction: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to extract medications: {str(e)}"
        )


@router.post("/anomaly-detection/{analysis_id}", response_model=AnomalyDetectionResponse)
async def detect_lab_anomalies(
    analysis_id: str,
    current_user: dict = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Detect anomalous lab results using Isolation Forest
    Identifies unusual patterns in medical metrics
    """
    try:
        # Get analysis
        analysis = db.query(Analysis).filter(
            Analysis.id == analysis_id
        ).first()
        
        if not analysis:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Analysis not found"
            )
        
        # Get all metrics for this analysis
        metrics = db.query(MedicalMetric).filter(
            MedicalMetric.analysis_id == analysis_id
        ).all()
        
        if not metrics:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="No metrics available for anomaly detection"
            )
        
        # Convert to dict format
        metrics_data = [
            {
                'metric_name': m.metric_name,
                'metric_value': m.metric_value,
                'reference_min': m.reference_min,
                'reference_max': m.reference_max,
                'unit': m.metric_unit,
                'status': m.status
            }
            for m in metrics
        ]
        
        # Detect anomalies
        anomalies = advanced_ml_service.detect_anomalies(metrics_data)
        
        return AnomalyDetectionResponse(
            analysis_id=analysis_id,
            anomalies=anomalies,
            anomaly_count=len(anomalies),
            total_metrics=len(metrics_data),
            message=f"Detected {len(anomalies)} anomalous results out of {len(metrics_data)} metrics"
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in anomaly detection: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to detect anomalies: {str(e)}"
        )


@router.get("/longitudinal-analysis", response_model=LongitudinalAnalysisResponse)
async def analyze_longitudinal_health(
    current_user: dict = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Analyze health trends over time using all user's analyses
    Tracks metrics longitudinally and predicts future trends
    """
    try:
        # Get all user's analyses ordered by date
        # Join with UploadedFile to filter by user_id
        analyses = db.query(Analysis).join(
            UploadedFile, Analysis.file_id == UploadedFile.id
        ).filter(
            UploadedFile.user_id == str(current_user['id'])
        ).order_by(Analysis.analysis_date.desc()).all()
        
        if len(analyses) < 2:
            return LongitudinalAnalysisResponse(
                user_id=str(current_user['id']),
                trends=[],
                predictions=[],
                risk_changes=[],
                insights=["Need at least 2 analyses for trend tracking. Upload more reports!"],
                data_points=len(analyses),
                tracking_period_days=0,
                message="Insufficient data for longitudinal analysis"
            )
        
        # Get current (latest) analysis metrics
        latest_analysis = analyses[0]
        current_metrics = db.query(MedicalMetric).filter(
            MedicalMetric.analysis_id == latest_analysis.id
        ).all()
        
        current_metrics_data = [
            {
                'metric_name': m.metric_name,
                'metric_value': m.metric_value,
                'reference_min': m.reference_min,
                'reference_max': m.reference_max,
                'unit': m.unit,
                'status': m.status
            }
            for m in current_metrics
        ]
        
        # Get historical analyses data
        historical_data = []
        for analysis in analyses[1:]:  # Skip latest (current)
            metrics = db.query(MedicalMetric).filter(
                MedicalMetric.analysis_id == analysis.id
            ).all()
            
            historical_data.append({
                'analysis_id': str(analysis.id),
                'analysis_date': analysis.analysis_date,
                'metrics': [
                    {
                        'metric_name': m.metric_name,
                        'metric_value': m.metric_value,
                        'status': m.status
                    }
                    for m in metrics
                ]
            })
        
        # Perform longitudinal analysis
        trend_results = advanced_ml_service.analyze_health_trends(
            current_metrics_data,
            historical_data
        )
        
        return LongitudinalAnalysisResponse(
            user_id=str(current_user['id']),
            trends=trend_results.get('trends', []),
            predictions=trend_results.get('predictions', []),
            risk_changes=trend_results.get('risk_changes', []),
            insights=trend_results.get('insights', []),
            data_points=trend_results.get('data_points', 0),
            tracking_period_days=trend_results.get('tracking_period_days', 0),
            message=f"Analyzed {len(analyses)} reports over {trend_results.get('tracking_period_days', 0)} days"
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in longitudinal analysis: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to perform longitudinal analysis: {str(e)}"
        )


@router.get("/health-check")
async def advanced_ml_health_check():
    """Check status of advanced ML features"""
    try:
        status = {
            'fine_tuned_biobert': advanced_ml_service.biobert_ner is not None,
            'medication_ner': advanced_ml_service.medication_ner is not None,
            'anomaly_detector': advanced_ml_service.anomaly_detector is not None,
            'longitudinal_tracking': True,
            'message': 'Advanced ML service is operational'
        }
        
        return status
        
    except Exception as e:
        logger.error(f"Error checking advanced ML health: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )
